File: backend/microservices/patient-service/src/main/java/com/hygie/patientservice/exception/GlobalExceptionHandler.java
package com.hygie.patientservice.exception;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.context.request.WebRequest;

import javax.validation.ConstraintViolationException;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;

/**
 * Gestionnaire global d'exceptions pour le service patient.
 *
 * Centralise la gestion des exceptions et fournit des réponses d'erreur cohérentes
 * pour toutes les APIs du service, conformément aux règles de développement Hygie-AI.
 *
 * @author Hygie-AI Team
 * @version 1.0.0
 */
@ControllerAdvice
public class GlobalExceptionHandler {

    private static final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);

    /**
     * Gère les exceptions liées aux patients non trouvés.
     *
     * @param ex L'exception levée
     * @param request La requête web
     * @return Une réponse d'erreur 404 avec des détails sur l'erreur
     */
    @ExceptionHandler(PatientNotFoundException.class)
    public ResponseEntity<ErrorResponse> handlePatientNotFoundException(
            PatientNotFoundException ex, WebRequest request) {
        // Assertion #1: Vérification que l'exception n'est pas null
        assert ex != null : "L'exception ne peut pas être null";

        // Log de l'erreur
        logger.warn("Patient non trouvé: {}", ex.getMessage());

        final ErrorResponse errorResponse = createErrorResponse(
                HttpStatus.NOT_FOUND,
                ex.getMessage(),
                request.getDescription(false));

        // Assertion #2: Vérification de la réponse
        assert errorResponse != null : "La réponse d'erreur ne peut pas être null";
        assert errorResponse.getStatus() == HttpStatus.NOT_FOUND.value() :
            "Le statut HTTP de la réponse d'erreur est incorrect";

        return new ResponseEntity<>(errorResponse, HttpStatus.NOT_FOUND);
    }

    /**
     * Gère les exceptions liées aux prescriptions non trouvées.
     *
     * @param ex L'exception levée
     * @param request La requête web
     * @return Une réponse d'erreur 404 avec des détails sur l'erreur
     */
    @ExceptionHandler(PrescriptionNotFoundException.class)
    public ResponseEntity<ErrorResponse> handlePrescriptionNotFoundException(
            PrescriptionNotFoundException ex, WebRequest request) {
        // Assertion #1: Vérification que l'exception n'est pas null
        assert ex != null : "L'exception ne peut pas être null";

        // Log de l'erreur
        logger.warn("Prescription non trouvée: {}", ex.getMessage());

        final ErrorResponse errorResponse = createErrorResponse(
                HttpStatus.NOT_FOUND,
                ex.getMessage(),
                request.getDescription(false));

        // Assertion #2: Vérification de la réponse
        assert errorResponse != null : "La réponse d'erreur ne peut pas être null";
        assert errorResponse.getStatus() == HttpStatus.NOT_FOUND.value() :
            "Le statut HTTP de la réponse d'erreur est incorrect";

        return new ResponseEntity<>(errorResponse, HttpStatus.NOT_FOUND);
    }

    /**
     * Gère les exceptions liées aux médicaments non trouvés.
     *
     * @param ex L'exception levée
     * @param request La requête web
     * @return Une réponse d'erreur 404 avec des détails sur l'erreur
     */
    @ExceptionHandler(MedicationNotFoundException.class)
    public ResponseEntity<ErrorResponse> handleMedicationNotFoundException(
            MedicationNotFoundException ex, WebRequest request) {
        // Assertion #1: Vérification que l'exception n'est pas null
        assert ex != null : "L'exception ne peut pas être null";

        // Log de l'erreur
        logger.warn("Médicament non trouvé: {}", ex.getMessage());

        final ErrorResponse errorResponse = createErrorResponse(
                HttpStatus.NOT_FOUND,
                ex.getMessage(),
                request.getDescription(false));

        // Assertion #2: Vérification de la réponse
        assert errorResponse != null : "La réponse d'erreur ne peut pas être null";
        assert errorResponse.getStatus() == HttpStatus.NOT_FOUND.value() :
            "Le statut HTTP de la réponse d'erreur est incorrect";

        return new ResponseEntity<>(errorResponse, HttpStatus.NOT_FOUND);
    }

    /**
     * Gère les exceptions liées aux violations de contraintes de validation.
     *
     * @param ex L'exception levée
     * @param request La requête web
     * @return Une réponse d'erreur 400 avec des détails sur les violations
     */
    @ExceptionHandler(ConstraintViolationException.class)
    public ResponseEntity<ValidationErrorResponse> handleConstraintViolationException(
            ConstraintViolationException ex, WebRequest request) {
        // Assertion #1: Vérification que l'exception n'est pas null
        assert ex != null : "L'exception ne peut pas être null";

        // Log de l'erreur
        logger.warn("Violation de contrainte: {}", ex.getMessage());

        final Map<String, String> errors = new HashMap<>();
        ex.getConstraintViolations().forEach(violation -> {
            String fieldName = violation.getPropertyPath().toString();
            String errorMessage = violation.getMessage();
            errors.put(fieldName, errorMessage);
        });

        final ValidationErrorResponse validationErrorResponse = new ValidationErrorResponse(
                HttpStatus.BAD_REQUEST.value(),
                "Erreur de validation",
                LocalDateTime.now(),
                request.getDescription(false),
                errors);

        // Assertion #2: Vérification de la réponse
        assert validationErrorResponse != null : "La réponse d'erreur de validation ne peut pas être null";
        assert validationErrorResponse.getStatus() == HttpStatus.BAD_REQUEST.value() :
            "Le statut HTTP de la réponse d'erreur est incorrect";
        assert !validationErrorResponse.getErrors().isEmpty() :
            "La liste des erreurs de validation ne peut pas être vide";

        return new ResponseEntity<>(validationErrorResponse, HttpStatus.BAD_REQUEST);
    }

    /**
     * Gère les exceptions liées aux arguments invalides dans les méthodes.
     *
     * @param ex L'exception levée
     * @param request La requête web
     * @return Une réponse d'erreur 400 avec des détails sur les erreurs de validation
     */
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ValidationErrorResponse> handleMethodArgumentNotValidException(
            MethodArgumentNotValidException ex, WebRequest request) {
        // Assertion #1: Vérification que l'exception n'est pas null
        assert ex != null : "L'exception ne peut pas être null";

        // Log de l'erreur
        logger.warn("Argument de méthode invalide: {}", ex.getMessage());

        final Map<String, String> errors = new HashMap<>();
        ex.getBindingResult().getAllErrors().forEach(error -> {
            String fieldName = ((FieldError) error).getField();
            String errorMessage = error.getDefaultMessage();
            errors.put(fieldName, errorMessage);
        });

        final ValidationErrorResponse validationErrorResponse = new ValidationErrorResponse(
                HttpStatus.BAD_REQUEST.value(),
                "Erreur de validation",
                LocalDateTime.now(),
                request.getDescription(false),
                errors);

        // Assertion #2: Vérification de la réponse
        assert validationErrorResponse != null : "La réponse d'erreur de validation ne peut pas être null";
        assert validationErrorResponse.getStatus() == HttpStatus.BAD_REQUEST.value() :
            "Le statut HTTP de la réponse d'erreur est incorrect";
        assert !validationErrorResponse.getErrors().isEmpty() :
            "La liste des erreurs de validation ne peut pas être vide";

        return new ResponseEntity<>(validationErrorResponse, HttpStatus.BAD_REQUEST);
    }

    /**
     * Gère les exceptions liées aux services Hygie.
     *
     * @param ex L'exception levée
     * @param request La requête web
     * @return Une réponse d'erreur avec des détails sur l'erreur de service
     */
    @ExceptionHandler(HygieServiceException.class)
    public ResponseEntity<ErrorResponse> handleHygieServiceException(
            HygieServiceException ex, WebRequest request) {
        // Assertion #1: Vérification que l'exception n'est pas null
        assert ex != null : "L'exception ne peut pas être null";

        // Log de l'erreur
        logger.error("Erreur de service Hygie: {}", ex.getMessage(), ex);

        final HttpStatus status = ex.getStatus() != null ? ex.getStatus() : HttpStatus.INTERNAL_SERVER_ERROR;

        final ErrorResponse errorResponse = createErrorResponse(
                status,
                ex.getMessage(),
                request.getDescription(false));

        // Assertion #2: Vérification de la réponse
        assert errorResponse != null : "La réponse d'erreur ne peut pas être null";
        assert errorResponse.getStatus() == status.value() :
            "Le statut HTTP de la réponse d'erreur est incorrect";

        return new ResponseEntity<>(errorResponse, status);
    }

    /**
     * Gère toutes les autres exceptions non spécifiquement traitées.
     *
     * @param ex L'exception levée
     * @param request La requête web
     * @return Une réponse d'erreur 500 avec des détails sur l'erreur
     */
    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleGlobalException(
            Exception ex, WebRequest request) {
        // Assertion #1: Vérification que l'exception n'est pas null
        assert ex != null : "L'exception ne peut pas être null";

        // Log de l'erreur
        logger.error("Erreur non gérée: {}", ex.getMessage(), ex);

        final ErrorResponse errorResponse = createErrorResponse(
                HttpStatus.INTERNAL_SERVER_ERROR,
                "Une erreur inattendue s'est produite. Veuillez contacter le support.",
                request.getDescription(false));

        // Assertion #2: Vérification de la réponse
        assert errorResponse != null : "La réponse d'erreur ne peut pas être null";
        assert errorResponse.getStatus() == HttpStatus.INTERNAL_SERVER_ERROR.value() :
            "Le statut HTTP de la réponse d'erreur est incorrect";

        return new ResponseEntity<>(errorResponse, HttpStatus.INTERNAL_SERVER_ERROR);
    }

    /**
     * Crée une réponse d'erreur standard.
     *
     * @param status Le statut HTTP
     * @param message Le message d'erreur
     * @param details Les détails de l'erreur
     * @return Une réponse d'erreur formatée
     */
    private ErrorResponse createErrorResponse(HttpStatus status, String message, String details) {
        // Assertion #1: Vérification des paramètres
        assert status != null : "Le statut HTTP ne peut pas être null";
        assert message != null && !message.isBlank() : "Le message d'erreur ne peut pas être null ou vide";

        final ErrorResponse errorResponse = new ErrorResponse(
                status.value(),
                message,
                LocalDateTime.now(),
                details);

        // Assertion #2: Vérification du résultat
        assert errorResponse != null : "La réponse d'erreur ne peut pas être null";
        assert errorResponse.getStatus() == status.value() :
            "Le statut HTTP de la réponse d'erreur est incorrect";

        return errorResponse;
    }
}

File: backend/microservices/llm-orchestrator/src/main/kotlin/com/hygie/llmorchestrator/exception/GlobalExceptionHandler.kt
package com.hygie.llmorchestrator.exception

import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.bind.support.WebExchangeBindException
import org.springframework.web.server.ServerWebExchange
import java.time.Instant
import javax.validation.ConstraintViolationException

/**
 * Représentation standardisée d'une erreur pour les réponses API.
 *
 * @property timestamp Horodatage de l'erreur
 * @property status Code HTTP de l'erreur
 * @property error Type d'erreur
 * @property message Message descriptif de l'erreur
 * @property path Chemin de la requête ayant provoqué l'erreur
 * @property errorCode Code d'erreur interne pour le suivi et la journalisation
 * @property details Détails supplémentaires sur l'erreur (facultatif)
 */
data class ApiError(
    val timestamp: Instant = Instant.now(),
    val status: Int,
    val error: String,
    val message: String,
    val path: String,
    val errorCode: String? = null,
    val details: List<String>? = null
)

/**
 * Gestionnaire global des exceptions pour les contrôleurs REST.
 *
 * Intercepte les exceptions non gérées et les convertit en réponses HTTP
 * formatées de manière cohérente.
 *
 * @author Hygie-AI Team
 */
@RestControllerAdvice
class GlobalExceptionHandler {
    private val logger = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)

    /**
     * Gère les exceptions spécifiques au service Hygie.
     *
     * @param ex L'exception levée
     * @param exchange L'échange web en cours
     * @return Une réponse formatée contenant les détails de l'erreur
     */
    @ExceptionHandler(HygieServiceException::class)
    fun handleHygieServiceException(
        ex: HygieServiceException,
        exchange: ServerWebExchange
    ): ResponseEntity<ApiError> {
        // Assertion #1: Vérification que le message d'erreur n'est pas vide
        val errorMessage = ex.message.takeIf { it.isNotBlank() } ?: "Erreur interne du service"

        // Journalisation de l'erreur
        logger.error("Erreur de service: {} (code: {})", errorMessage, ex.errorCode, ex)

        // Construction de la réponse d'erreur
        val apiError = ApiError(
            status = ex.httpStatus.value(),
            error = ex.httpStatus.reasonPhrase,
            message = errorMessage,
            path = exchange.request.path.value(),
            errorCode = ex.errorCode,
            details = ex.cause?.message?.let { listOf(it) }
        )

        // Assertion #2: Vérification que le statut HTTP est cohérent
        require(ex.httpStatus.value() >= 400) { "Le statut HTTP doit être un code d'erreur (>= 400)" }

        return ResponseEntity
            .status(ex.httpStatus)
            .body(apiError)
    }

    /**
     * Gère les exceptions de validation des contraintes.
     *
     * @param ex L'exception de validation
     * @param exchange L'échange web en cours
     * @return Une réponse formatée contenant les détails des violations
     */
    @ExceptionHandler(ConstraintViolationException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handleConstraintViolationException(
        ex: ConstraintViolationException,
        exchange: ServerWebExchange
    ): ResponseEntity<ApiError> {
        logger.warn("Erreur de validation des contraintes: {}", ex.message, ex)

        // Extraction des détails des violations
        val details = ex.constraintViolations.map { violation ->
            "${violation.propertyPath}: ${violation.message}"
        }

        val apiError = ApiError(
            status = HttpStatus.BAD_REQUEST.value(),
            error = HttpStatus.BAD_REQUEST.reasonPhrase,
            message = "Erreur de validation des données",
            path = exchange.request.path.value(),
            errorCode = "HYGIE-VALIDATION-ERROR",
            details = details
        )

        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(apiError)
    }

    /**
     * Gère les exceptions de liaison des échanges web.
     *
     * @param ex L'exception de liaison
     * @param exchange L'échange web en cours
     * @return Une réponse formatée contenant les détails des erreurs de liaison
     */
    @ExceptionHandler(WebExchangeBindException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    fun handleWebExchangeBindException(
        ex: WebExchangeBindException,
        exchange: ServerWebExchange
    ): ResponseEntity<ApiError> {
        logger.warn("Erreur de liaison des données: {}", ex.message, ex)

        // Extraction des détails des erreurs de champ
        val details = ex.bindingResult.fieldErrors.map { fieldError ->
            "${fieldError.field}: ${fieldError.defaultMessage}"
        }

        val apiError = ApiError(
            status = HttpStatus.BAD_REQUEST.value(),
            error = HttpStatus.BAD_REQUEST.reasonPhrase,
            message = "Erreur de validation des données de requête",
            path = exchange.request.path.value(),
            errorCode = "HYGIE-REQUEST-VALIDATION-ERROR",
            details = details
        )

        return ResponseEntity
            .status(HttpStatus.BAD_REQUEST)
            .body(apiError)
    }

    /**
     * Gère toutes les autres exceptions non spécifiquement traitées.
     *
     * @param ex L'exception non gérée
     * @param exchange L'échange web en cours
     * @return Une réponse formatée générique
     */
    @ExceptionHandler(Exception::class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    fun handleGenericException(
        ex: Exception,
        exchange: ServerWebExchange
    ): ResponseEntity<ApiError> {
        logger.error("Exception non gérée: {}", ex.message, ex)

        val apiError = ApiError(
            status = HttpStatus.INTERNAL_SERVER_ERROR.value(),
            error = HttpStatus.INTERNAL_SERVER_ERROR.reasonPhrase,
            message = "Une erreur interne est survenue",
            path = exchange.request.path.value(),
            errorCode = "HYGIE-INTERNAL-ERROR"
        )

        return ResponseEntity
            .status(HttpStatus.INTERNAL_SERVER_ERROR)
            .body(apiError)
    }
}
